### PR TITLE
fix: allow `u0` to be duals

### PR DESCRIPTION
**Pull request by avik-pal**

### CODE REVIEW

File: src/internal/forward_diff.jl [hunk 1]
```diff
@@ -2,16 +2,33 @@
...
-        prob::NonlinearProblem{<:Union{Number, <:AbstractArray}, iip,
-            <:Union{<:Dual{T, V, P}, <:AbstractArray{<:Dual{T, V, P}}}},
-        alg::Union{Nothing, AbstractNonlinearAlgorithm},
-        args...;
-        kwargs...) where {T, V, P, iip}
-    sol, partials = __nlsolve_ad(prob, alg, args...; kwargs...)
-    dual_soln = __nlsolve_dual_soln(sol.u, partials, prob.p)
-    return SciMLBase.build_solution(
-        prob, alg, dual_soln, sol.resid; sol.retcode, sol.stats, sol.original)
+for (uType, pType) in [
```
Comment (oscardssmith) [inline src/internal/forward_diff.jl:5]: Why is this necessary? This seems pretty ugly

Comment (avik-pal) [reply src/internal/forward_diff.jl:5]: ambiguity prevention

Comment (oscardssmith) [reply src/internal/forward_diff.jl:5]: is there a reason this doesn't overload `__init` and `__solve`  like `OrdinaryDiffEq does?

Comment (avik-pal) [reply src/internal/forward_diff.jl:5]: I was asking myself the same question today morning

Comment (ChrisRackauckas) [reply src/internal/forward_diff.jl:5]: i think you need to so that you promote first.

**Comment by avik-pal** [comment]
This is now handled in #458 (via dispatches on __init & __solve)
